File: as-2020-api-master/AS2020Api/Model/EventStatusScoreCalculator.cs
using System;

namespace AS2020Api.Model
{
    public class EventStatusScoreCalculator
    {
        private readonly GameParameters _parameters;

        public EventStatusScoreCalculator(GameParameters parameters)
        {
            _parameters = parameters;
        }

        public int CalculateScoreForEvent(PlayEvent playEvent, GameStatus status)
        {
            var properties = PlayEventProperties.GetByType(playEvent.Type);
            var baseCost = properties.BaseEventScore;
            var stepCount = properties.StepCount(status);
            var stepRatio = 1 + ((float)_parameters.ItemStepRate / 100);
            var stepTotalCost = (int) Math.Ceiling(Math.Pow(stepRatio, stepCount) * baseCost);

            return stepTotalCost;
        }
        
        public int CalculateScoreForTick(PlayEvent playEvent, GameStatus status)
        {
            var properties = PlayEventProperties.GetByType(playEvent.Type);
            return properties.BaseTickScore;
        }
    }
}

File: as-2020-ui-master/src/model/event-status-score-calculator.ts
import { GameParameters } from './game-parameters';
import { PlayEventType } from './play-event';
import { GameStatus } from './game-status';
import { getPropertiesByType } from './play-event-properties';
import { getAdditionalPlayEventScore } from './player-status';

export interface EventStatusScoreCalculator {
  calculateScoreForEvent: (e: PlayEventType, gameStatus: GameStatus) => number;
  calculateScoreForTick: (e: PlayEventType, gameStatus: GameStatus) => number;
}

export const EventStatusScoreCalculator = (
  parameters: GameParameters
): EventStatusScoreCalculator => {
  const calculateScoreForEvent = (
    e: PlayEventType,
    gameStatus: GameStatus
  ): number => {
    const baseCost = parameters.baseScoreRates[e.name];
    const properties = getPropertiesByType(e);
    const stepCount = properties.stepCount;
    const stepRatio = 1 + parameters.itemStepRate / 100;
    const stepTotalCost = Math.ceil(
      Math.pow(stepRatio, stepCount(gameStatus)) * baseCost
    );
    const additionalScore = getAdditionalPlayEventScore(
      gameStatus.playerStatus,
      e,
      gameStatus
    );
    return Math.floor(stepTotalCost + additionalScore);
  };

  const calculateScoreForTick = (e: PlayEventType) =>
    parameters.baseScorePerTick[e.name];

  return {
    calculateScoreForEvent,
    calculateScoreForTick
  };
};
